### DISCUSSION TITLE
Micronaut Data 4 / JPA question

### DISCUSSION THREAD

**Discussion by ethanhere** [micronaut-projects/micronaut-data#2803] [Q&A]
One of our micronaut 3 application has application.yml:
jpa:
  default:
    packages-to-scan:
      - 'myapp.entities'
 
With micronaut 4 I got an test error:
[ERROR]   DaoTest » BeanInstantiation Bean definition [org.hibernate.SessionFactory] could not be loaded: Error instantiating bean of type  [org.hibernate.boot.SessionFactoryBuilder]
 
Message: Entities not found for JPA configuration: 'default' within packages ['myapp.entities']. Check that you have correctly specified a package containing JPA entities within the "jpa.default.entity-scan.packages" property in your application configuration and that those entities are either compiled with Micronaut or a build time index produced with @Introspected(packages="foo.bar", includedAnnotations=Entity.class) declared on your Application class
Path Taken: SessionFactoryPerDataSourceFactory.buildHibernateSessionFactoryBuilder(SessionFactoryBuilder sessionFactoryBuilder) --> SessionFactoryPerDataSourceFactory.buildHibernateSessionFactoryBuilder([SessionFactoryBuilder sessionFactoryBuilder]) --> SessionFactoryPerDataSourceFactory.buildHibernateSessionFactoryBuilder([Metadata metadata],JpaConfiguration jpaConfiguration,String name)
 
So according to https://micronaut-projects.github.io/micronaut-data/4.6.0/guide/index.html
application configuration file should look like
jpa:
  default:
    entity-scan:
        packages: 'myapp.entities'
 
In the project there is a @Entity class located in package 'myapp.entities', so I changed application.yml to
jpa:
  default:
    entity-scan:
      packages: 'myapp.entities'

I also changed my @MicronautTest code to
@MicronautTest
@Introspected(packages="myapp.entities", includedAnnotations=Entity.class)
public class DaoTest {
...

 
But it still doesn’t work, and the error message is the same. How can I fix that? Do you have any detailed documentation, especially explaining how to migrate micronaut data 3 to 4?
 
Thanks,
Ethan

**Comment by radovanradic** [comment]
It would help if you can share your example project, something might be missing in the build file.

**Reply by ethanhere** [reply]
@radovanradic thanks for responding. it's a bit complicated to share my company's source code. how can I generate a Micronaut Data 4 project with hibernate jpa and see if I can reproduce it?

Another question is that, in the long error message, I may not understand accurately on "a build time index produced with @introspected(packages="foo.bar", includedAnnotations=Entity.class) declared on your Application class" part.
In my original post, I added @introspected annotation to @MicronautTest class:
@MicronautTest
@introspected(packages="myapp.entities", includedAnnotations=Entity.class)
public class DaoTest {

Is that the right syntax to trigger micronaut 4 annotation processor to scan javax.persistence.Entity.class under myapp.entities package?


According to https://micronaut-projects.github.io/micronaut-liquibase/latest/guide/, at 4 Configuration section, it suggests to have application.yml configured as:
jpa:
    default:
        packages-to-scan:
            - 'example.micronaut'
        properties:
            hibernate:
                hbm2ddl:
                    auto: none
                show_sql: true

However, the error message references configuration jpa.default.entity-scan.packages (this also appears in https://micronaut-projects.github.io/micronaut-data/4.6.0/guide/index.html). Can someone clarify what should be the right configuration? Or both are correct but different usages?

**Comment by radovanradic** [comment]
Try Micronaut Launch, this [link](https://micronaut.io/launch?type=DEFAULT&name=demo&package=com.example&javaVersion=JDK_17&lang=JAVA&build=GRADLE&test=JUNIT&features=h2&features=data-jpa&version=4.3.2) will have predefined data-jpa and h2 feature for simple use case.
Adding `@Entity` should be enough, if you can try fresh example from the Micronaut Launch then it should work.

**Reply by sbscomp** [reply]
@radovanradic I've got the exact same problem as @ethanhere, but a project that's basic enough I was able to create a sample project. I created a sample project in Micronaut Launch, manually created a minimal scenario that matches my project, and it works. Copied over my controller, repository, entity, and everything works exactly as expected. I pared down my work project all the way down so it's an exact replica of the micronaut launch project, and I get the introspection error as Ethan does. The _only_ difference between the two at this point is that my work project was generated by IntelliJ with Java 17 and Micronaut 4.3.0 as the original configuration, and I manually updated to Java 21 & Micronaut 4.3.2 by manually changing the build.gradle.kts file. I have absolutely no clue why that would be a problem, but it is, and it's enough to drive someone bonkers. 

FWIW, I re-generated a new project with 4.3.2 from IntelliJ, and copied everything over, and it also works as expected.

I'm happy to share the two projects privately if it helps to identify where this issue is coming from, but whatever is going on here, it feels like there's an issue with any scenario where the project started from an earlier version and has been updated to 4.3.2 by changing the gradle build files, in such a way that it is not readily identifiable to a casual observation. If that's truly the case, this is going to end up being a massive problem for any legacy projects that want to upgrade to 4.3.2.

**Reply by ethanhere** [reply]
I solved the problem by changing javax.persistence.Entity to jakarta.persistence.Entity. In fact I changed every javax.persistence.* to jakarta.persistence.*; dependency wise changed  <artifactId>javax.persistence-api</artifactId> to      <artifactId>jakarta.persistence-api</artifactId>
I also added
        <dependency>
            <groupId>io.micronaut.data</groupId>
            <artifactId>micronaut-data-tx-hibernate</artifactId>
        </dependency>
After this change the original issue is resolved. In that maven module, the first test passed. The second test still complains:
Message: No bean of type [jakarta.persistence.EntityManager] exists. 

Where can I find jakarta.persistence.EntityManager related artifacts?

**Comment by radovanradic** [comment]
Do you have dependency `io.micronaut.sql:micronaut-jdbc-hikari`
It's referred in this [guide](https://guides.micronaut.io/latest/micronaut-jpa-hibernate-maven-java.html), maybe that guide could be helpful.

**Reply by ethanhere** [reply]
I have.

**Reply by ethanhere** [reply]
Hi @radovanradic 

Thanks for the reference link. I checked and compared the sample app https://guides.micronaut.io/latest/micronaut-jpa-hibernate-maven-java.zip with my app:
I have all dependencies specified in 4.1: 
<dependency> 
    <groupId>io.micronaut.sql</groupId>
    <artifactId>micronaut-hibernate-jpa</artifactId>
    <scope>compile</scope>
</dependency>
<dependency> 
    <groupId>io.micronaut.data</groupId>
    <artifactId>micronaut-data-tx-hibernate</artifactId>
    <scope>compile</scope>
</dependency>
<dependency> 
    <groupId>io.micronaut.sql</groupId>
    <artifactId>micronaut-jdbc-hikari</artifactId>
    <scope>compile</scope>
</dependency>
<dependency> 
    <groupId>com.h2database</groupId>
    <artifactId>h2</artifactId>
    <scope>runtime</scope>
</dependency>

I have application.yml rather than application.properties:
datasources:
  default:
    url: jdbc:h2:mem:default;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE
    username: use
    password: ""
    driverClassName: org.h2.Driver

jpa:
  default:
    packages-to-scan:
      - 'myapp.entities'
    properties:
      hibernate:
        hbm2ddl:
          auto: validate
        show_sql: true
        format_sql: true

Java code:
import jakarta.persistence.EntityManager;
import jakarta.persistence.EntityManagerFactory;
import jakarta.persistence.Persistence;
import jakarta.persistence.PersistenceContext;
import jakarta.persistence.TypedQuery;

import lombok.extern.slf4j.Slf4j;

import static java.util.stream.Collectors.toList;

@Slf4j
@Singleton
public class ServiceDao {

//    @PersistenceContext
    private EntityManager db;

    private final ObjectMapper mapper;
    private final FrameworkClockProvider clockProvider;

    @Inject
    public AliServiceDao(ObjectMapper mapper, FrameworkClockProvider clockProvider, EntityManager db) {

        this.mapper = mapper;
        this.clockProvider = clockProvider;
//        EntityManagerFactory entityManagerFactory = Persistence.createEntityManagerFactory("default");
//        this.db = entityManagerFactory.createEntityManager();
        this.db = db;
    }

When running test
@MicronautTest
public class ServiceDaoTest {
    @Inject
    private ServiceDao dao;


I got error:
Caused by: io.micronaut.context.exceptions.NoSuchBeanException: No bean of type [jakarta.persistence.EntityManager] exists. 
	at io.micronaut.context.DefaultBeanContext.newNoSuchBeanException(DefaultBeanContext.java:2787)

Can you suggest how to fix this? I don't see I miss any dependency comparing with sample application and the guide (https://guides.micronaut.io/latest/micronaut-jpa-hibernate-maven-java.html)

**Reply by radovanradic** [reply]
It would be the best if you can post minimal reproducer app so we might easier figure out what is wrong.

**Reply by ethanhere** [reply]
[micronaut4xp-jpa.zip](https://github.com/micronaut-projects/micronaut-data/files/14407933/micronaut4xp-jpa.zip)
Hi @radovanradic 
attached is a maven profile of repro. running command "mvn clean install" from the extracted directory (micronaut4xp-jpa) would get the error. In the zip file, there is a file build.txt, that's the output of command "mvn clean install" as well.

Error:
[ERROR] myapp.ServiceDaoTest.testCreateGetNew -- Time elapsed: 0.143 s <<< ERROR!
io.micronaut.context.exceptions.DependencyInjectionException: 
Failed to inject value for parameter [db] of class: myapp.$ServiceDao$Definition$Intercepted

Message: No bean of type [jakarta.persistence.EntityManager] exists. 
* [EntityManager] requires the presence of a bean of type [org.hibernate.SessionFactory].
 * [SessionFactory] requires the presence of a bean of type [org.hibernate.boot.SessionFactoryBuilder].
  * [SessionFactoryBuilder] requires the presence of a bean of type [org.hibernate.boot.Metadata].
   * [Metadata] requires the presence of a bean of type [org.hibernate.boot.MetadataSources].
    * [MetadataSources] requires the presence of a bean of type [org.hibernate.service.ServiceRegistry].
     * [ServiceRegistry] requires the presence of a bean of type [javax.sql.DataSource].
      * [DataSource] requires the presence of a bean of type [io.micronaut.configuration.jdbc.hikari.DatasourceConfiguration].
       * [DatasourceConfiguration] is disabled because:
        - Configuration requires entries under the prefix: [datasources.default]
     * [ServiceRegistry] is disabled because:
      - Existing bean [io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory] of type [io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory] registered in context
    * [MetadataSources] is disabled because:
     - Existing bean [io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory] of type [io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory] registered in context
   * [Metadata] is disabled because:
    - Existing bean [io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory] of type [io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory] registered in context
  * [SessionFactoryBuilder] is disabled because:
   - Existing bean [io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory] of type [io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory] registered in context
 * [SessionFactory] is disabled because:
  - Existing bean [io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory] of type [io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory] registered in context
Path Taken: ServiceDaoTest.dao --> new $ServiceDao$Definition$Intercepted(ObjectMapper mapper,[EntityManager db],BeanResolutionContext $beanResolutionContext,BeanContext $beanContext,Qualifier $qualifier,List $interceptors,InterceptorRegistry $interceptorRegistry)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2181)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForConstructorArgument(AbstractInitializableBeanDefinition.java:1328)
	at myapp.$ServiceDao$Definition$Intercepted$Definition.instantiate(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2330)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2300)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2312)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3114)
	at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:80)
	at io.micronaut.context.DefaultBeanContext.findOrCreateSingletonBeanRegistration(DefaultBeanContext.java:3016)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2977)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2751)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1750)
	at io.micronaut.context.AbstractBeanResolutionContext.getBean(AbstractBeanResolutionContext.java:89)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2165)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForField(AbstractInitializableBeanDefinition.java:1672)
	at myapp.$ServiceDaoTest$Definition.inject(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doInjectAndInitialize(DefaultBeanContext.java:2657)
	at io.micronaut.context.DefaultBeanContext.inject(DefaultBeanContext.java:1025)
	at io.micronaut.test.extensions.AbstractMicronautExtension.beforeEach(AbstractMicronautExtension.java:454)
	at io.micronaut.test.extensions.junit5.MicronautJunit5Extension.beforeEach(MicronautJunit5Extension.java:223)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: io.micronaut.context.exceptions.NoSuchBeanException: No bean of type [jakarta.persistence.EntityManager] exists. 
* [EntityManager] requires the presence of a bean of type [org.hibernate.SessionFactory].
 * [SessionFactory] requires the presence of a bean of type [org.hibernate.boot.SessionFactoryBuilder].
  * [SessionFactoryBuilder] requires the presence of a bean of type [org.hibernate.boot.Metadata].
   * [Metadata] requires the presence of a bean of type [org.hibernate.boot.MetadataSources].
    * [MetadataSources] requires the presence of a bean of type [org.hibernate.service.ServiceRegistry].
     * [ServiceRegistry] requires the presence of a bean of type [javax.sql.DataSource].
      * [DataSource] requires the presence of a bean of type [io.micronaut.configuration.jdbc.hikari.DatasourceConfiguration].
       * [DatasourceConfiguration] is disabled because:
        - Configuration requires entries under the prefix: [datasources.default]
     * [ServiceRegistry] is disabled because:
      - Existing bean [io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory] of type [io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory] registered in context
    * [MetadataSources] is disabled because:
     - Existing bean [io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory] of type [io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory] registered in context
   * [Metadata] is disabled because:
    - Existing bean [io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory] of type [io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory] registered in context
  * [SessionFactoryBuilder] is disabled because:
   - Existing bean [io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory] of type [io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory] registered in context
 * [SessionFactory] is disabled because:
  - Existing bean [io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory] of type [io.micronaut.configuration.hibernate.jpa.conf.SessionFactoryPerDataSourceFactory] registered in context
	at io.micronaut.context.DefaultBeanContext.newNoSuchBeanException(DefaultBeanContext.java:2787)
	at io.micronaut.context.DefaultApplicationContext.newNoSuchBeanException(DefaultApplicationContext.java:307)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2756)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1750)
	at io.micronaut.context.AbstractBeanResolutionContext.getBean(AbstractBeanResolutionContext.java:89)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2165)
	... 21 more

**Reply by radovanradic** [reply]
I have added this dependency to pom.xml (because in Micronaut 4 application.properties is default file and for yml need to include `org.yaml:snakeyaml` dependency
```
<dependency>
            <groupId>org.yaml</groupId>
            <artifactId>snakeyaml</artifactId>
            <scope>runtime</scope>
        </dependency>
```
and also updated this in jpa configuration (changed `jpa.default.properties.hibernate.hbm2ddl.auto` value from `validate` to `update` because there were no tables created and test needs them)
```
jpa:
  default:
    packages-to-scan:
      - 'myapp.entities'
    properties:
      hibernate:
        hbm2ddl:
          auto: update
        show_sql: true
        format_sql: true
```

Hope this helps.

**Reply by ethanhere** [reply]
Thanks @radovanradic . what you suggested does fix the repro project. However adding snakeyaml dependency and change jpa.default.properties.hibernate.hbm2ddl.auto value from validate to update, still doesn't fix my work project. I am getting the following error message:
[ERROR] myapp.ServiceDaoTest.testCreate -- Time elapsed: 0.001 s <<< ERROR!
io.micronaut.context.exceptions.DependencyInjectionException: 
Failed to inject value for parameter [db] of class: myapp.$ServiceDao$Definition$Intercepted

Message: No bean of type [jakarta.persistence.EntityManager] exists. 
Path Taken: ServiceDaoTest.dao --> new $ServiceDao$Definition$Intercepted(ObjectMapper mapper,[EntityManager db],BeanResolutionContext $beanResolutionContext,BeanContext $beanContext,Qualifier $qualifier,List $interceptors,InterceptorRegistry $interceptorRegistry)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2181)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForConstructorArgument(AbstractInitializableBeanDefinition.java:1328)
	at myapp.$ServiceDao$Definition$Intercepted$Definition.instantiate(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2330)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2300)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2312)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:3114)
	at io.micronaut.context.SingletonScope.getOrCreate(SingletonScope.java:80)
	at io.micronaut.context.DefaultBeanContext.findOrCreateSingletonBeanRegistration(DefaultBeanContext.java:3016)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2977)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2751)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1750)
	at io.micronaut.context.AbstractBeanResolutionContext.getBean(AbstractBeanResolutionContext.java:89)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2165)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForField(AbstractInitializableBeanDefinition.java:1672)
	at com.signify.iot.platform.security.serviceidm.data.$AliServiceDaoTest$Definition.inject(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doInjectAndInitialize(DefaultBeanContext.java:2657)
	at io.micronaut.context.DefaultBeanContext.inject(DefaultBeanContext.java:1025)
	at io.micronaut.test.extensions.AbstractMicronautExtension.beforeEach(AbstractMicronautExtension.java:454)
	at io.micronaut.test.extensions.junit5.MicronautJunit5Extension.beforeEach(MicronautJunit5Extension.java:223)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: io.micronaut.context.exceptions.NoSuchBeanException: No bean of type [jakarta.persistence.EntityManager] exists. 
	at io.micronaut.context.DefaultBeanContext.newNoSuchBeanException(DefaultBeanContext.java:2787)
	at io.micronaut.context.DefaultApplicationContext.newNoSuchBeanException(DefaultApplicationContext.java:307)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2756)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1750)
	at io.micronaut.context.AbstractBeanResolutionContext.getBean(AbstractBeanResolutionContext.java:89)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:2165)
	... 21 more

comparing this error with repro pro's error, this error message simply says "No bean of type [jakarta.persistence.EntityManager] exists" but doesn't gave more details. (I have added snakeyaml dependency). I also checked all dependencies included in my work project.

**Reply by radovanradic** [reply]
Sorry, not able to tell what is wrong without looking at the project.

**Reply by ethanhere** [reply]
[micronaut4xp-entitymanager-nobean.zip](https://github.com/micronaut-projects/micronaut-data/files/14411711/micronaut4xp-entitymanager-nobean.zip)
Hi @radovanradic 
Attached is a project that I stripped as much as from my work project and preserve the " No bean of type [jakarta.persistence.EntityManager] exists" error. Can you please take a look? I checked the project with the guide on dependencies, etc. but it still keeps getting " No bean of type [jakarta.persistence.EntityManager] exists" error

**Reply by radovanradic** [reply]
Mn4 uses hibernate6. Try changing this hibernate dependency in Maven
```
<dependency>
            <groupId>org.hibernate.orm</groupId>
            <artifactId>hibernate-core</artifactId>
            <version>6.4.4.Final</version>
            <exclusions>
                <exclusion>
                    <groupId>org.jboss.logging</groupId>
                    <artifactId>jboss-logging</artifactId>
                </exclusion>
                <exclusion>
                    <groupId>org.glassfish.jaxb</groupId>
                    <artifactId>jaxb-runtime</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
```
Had errors without annotations and bytebuddy so I removed those exclusions, not sure if that will be an issue for you.

**Reply by ethanhere** [reply]
Thanks @radovanradic , that works!
just curious, how did you find that out? with the version (5.6.15.Final), was the EntityManager implementation missing?

**Reply by radovanradic** [reply]
I have seen javax.persistence in the dependencies and then noticed 5.6.15 org.hibernate and figured out it should switch to hibernate6.

**Reply by ethanhere** [reply]
thanks
